feat(yarn2nix): arg parsing using optparse-applicative

sternenseemann opened the pull request:
This adds backwards compatible arg parsing via `optparse-applicative` which can be used for both actions. It has slightly less usability to reduced quality in help messages, but should be more extensible.

Main motivation is creating the possibility to add new flags (for another feature for `--template`) without increasing the mess in `cli`.

---

```
Unify runtime "configuration" mechanism: We now generate a RunConfig
from the command line and pass it to cli. This config is used to
determine which action to take (template / lock file) via RunMode
and contains additional information to influence runtime behavior
like runOffline.

The configuration can be passed to both actions if necessary, mandatory
action specific flags / variables could be hidden away in RunMode in the
future.
```

sternenseemann committed:
feat(yarn2nix): arg parsing using optparse-applicative

Unify runtime "configuration" mechanism: We now generate a RunConfig
from the command line and pass it to cli. This config is used to
determine which action to take (template / lock file) via RunMode
and contains additional information to influence runtime behavior
like runOffline.

The configuration can be passed to both actions if necessary, mandatory
action specific flags / variables could be hidden away in RunMode in the
future.

sternenseemann committed:
chore(yarn2nix.nix): update via cabal2nix

Minus references to hpack since this seems to not work with the used
version of nixpkgs.

Profpatsch commented:
What does the new help text look like?

sternenseemann commented:
Could be much better, but I fear we need to rework the CLI completely for things to improve

```
Usage: yarn2nix [--template] [--offline] [FILE]
  yarn2nix has two modes: In its default mode (started without --template) it
  parses a given yarn.lock file and prints a nix expressions representing it to
  stdout. If --template is given, it processes a given package.json and prints a
  template nix expression for an equivalent nix package. In both modes yarn2nix
  will take the file as an argument or read it from stdin if it is missing.

Available options:
  --template               Output a nix package template for a given
                           package.json
  --offline                Makes yarn2nix fail if network access is required
  -h,--help                Show this help text
```

Profpatsch committed:
fix(Main): improve the usage text linebreaks

Profpatsch commented:
Slightly dumb, but now the output looks like this:

```
Usage: yarn2nix [--template] [--offline] [FILE]
  yarn2nix has two modes:
  
  In its default mode (started without --template) it parses a given yarn.lock file
  and prints a nix expressions representing it to stdout.
  
  If --template is given, it processes a given package.json
  and prints a template nix expression for an equivalent nix package.
  
  In both modes yarn2nix will take the file as an argument
  or read it from stdin if it is missing.

Available options:
  --template               Output a nix package template for a given
                           package.json
  --offline                Makes yarn2nix fail if network access is required
  -h,--help                Show this help text
```

Profpatsch commented on package.yaml:
```diff
@@ -50,6 +50,7 @@ executables:
     main: Main.hs
     dependencies:
       - yarn2nix
+      - optparse-applicative >= 0.13 && < 0.15
```
Can leave out upper boundary, unnecessary

Profpatsch commented on src/Distribution/Nixpkgs/Nodejs/Cli.hs:
```diff
@@ -87,12 +67,9 @@ cli = \case
 
 die' :: Text -> IO a
 die' err = putErrText err *> exitFailure
-dieWithUsage :: Text -> IO ()
-dieWithUsage err = die' (err <> "\n" <> usage)
```
What happened to `dieWithUsage`?

sternenseemann replied:
I don't think it's possible to print the optparse help text there and I didn't want to duplicate it. It might be possible to extract the usage from a `ParserInfo` or similar, I'm not sure.

Profpatsch replied:
Should work if you use

```
parserUsage :: ParserPrefs -> Parser a -> String -> Doc
```

sternenseemann commented:
```
  --template               Output a nix package template for a given
                           package.json
```

I'm annoyed with `optparse-applicative`'s arbitrary line wrapping not sure if it can be configured.

Profpatsch commented:
It’s strange, because it doesn’t arbitrarily break the doc, which is why I had to add the `linebreak` calls …

sternenseemann committed:
chore(package.yaml): relax bounds for optparse-applicative

sternenseemann committed:
feat(yarn2nix): return of dieWithUsage

Use parserFailure and renderFailure to obtain the help text.
optparse-applicative changes the constructor of ShowHelpText so we
introduce an upper constraint.

sternenseemann commented:
Managed to fix the odd linebreak and return `dieWithUsage` in 4eaf77fd11faaec54a757b84d657bd9b630d7cfe. `Cli.hs` is bit of a mess now, I feel like, feedback appreciated.

Profpatsch commented:
whoo \oooo/
